docs/rfc/rfc-8.adoc

```diff
@@ -0,0 +1,410 @@
+
+==== Wormhole
+
+The Wormhole bridging process from L1 to L2 (or sidechain) will look as follows:
+
+* There is a tBTC holder on L1. The holder goes to the Wormhole Portal and
+  selects the chain they want to bridge to.
+* The holder submits one transaction to L1. After the transaction is mined, they
+  wait about 15 minutes for the Ethereum block finality.
+* The holder submits one transaction to L2. After that transaction is mined,
+  they have their tBTC on L2.
+
+The Wormhole bridging process from L2 (or sidechain) to L1 will look as follows:
+
+* There is a tBTC holder on L2. That holder goes to the Wormhole portal and
+  selects one of the L2 chains they want to bridge from.
+* The holder submits one transaction to L2. After the transaction is mined, they
+  wait about 15 minutes for the L2 block finality.
+* The holder submits one transaction to L1. After that transaction is mined,
+  they have their tBTC on L1.
+
+What is not immediately obvious is that the token received on L2 may not be the
+canonical token on that L2.
+
+For example, when I crossed the bridge with ETH, I received WETH token at
+address link:https://arbiscan.io/token/0xd8369c2eda18dd6518eabb1f85bd60606deb39ec[`0xd8369c2eda18dd6518eabb1f85bd60606deb39ec`].
+This is not the canonical WETH on Arbitrum. The canonical WETH on Arbitrum is
+link:https://arbiscan.io/token/0x82af49447d8a07e3bd95bd0d56f35241523fbab1[`0x82af49447d8a07e3bd95bd0d56f35241523fbab1`].
+When I crossed the bridge with USDC I received USDC token at
+link:https://arbiscan.io/token/0xc96f2715e2a242d50d1b0bc923dbe1740b8ecf18[`0xc96f2715e2a242d50d1b0bc923dbe1740b8ecf18`]
+which is not the canonical USDC on Arbitrum. The canonical USDC on Arbitrum is
+link:https://arbiscan.io/token/0xff970a61a04b1ca14834a43f5de4533ebddb5cc8[`0xff970a61a04b1ca14834a43f5de4533ebddb5cc8`].
+
+Wormhole cannot magically mint all tokens on Arbitrum or other L2s. Wormhole
+mints its own token representing the bridged asset that may be swapped to the
+canonical representation on the given L2 assuming there is a market for such
+a swap. The overview of liquid markets on each target chain is available in the
+link:https://docs.wormhole.com/wormhole/overview-liquid-markets[Wormhole documentation].
+Not every chain and token pair has a market. 
+
+To fully use the capabilities of the Wormhole portal and make the user
+experience seamless, we will implement a gateway contract having the authority to
+wrap Wormhole-specific tokens into tBTC tokens in a 1:1 ratio. This contract
+will have a minting authority for tBTC on L2. This way, no liquid market has to
+exist on any target chain for users to be able to cross the Wormhole bridge into
+the canonical tBTC.
+
+IMPORTANT: The requirement for a seamless user experience is that the Wormhole
+team has to integrate the step of wrapping the token using the gateway contract
+into the Wormhole bridge flow, in the same transaction as the L2 confirmation.
```

Comment:
How can this be done in the same transaction?  I believe it would need to be done in two transactions.

Option 1:
a. First, redeem "whtBTC" on L2 address using Wormhole token contract on L2.
b. Second, redeem whtBTC for tBTC on L2 using canonical token contract on L2.

Option 1:
a. First, give approval to canonical token contract to spend "whtBTC" for user on L2.
b. Second, redeem whtBTC onto the canonical token contract which can then atomically swap for tBTC.

Reply:
That really depends on the bridge. We wrote this assuming there would be a way for user with address A on L1 to bridge to address B on the sidechain or L2.

If the existing bridging infra has to go from address A to address A between chains, then this wrapping idea isn't helpful, and the bridge should instead directly mint against the L2 tBTC contract.

Reply:
> Maybe there's another way I'm missing?

This would require custom handling in the bridge.

I think there are three ways available:
1. Update the L2 Wormhole contract to allow for an optional wrapping into a canonical token in the L2 redemption transaction. Under the hood, the Wormhole contract would mint whtBTC and wrap it using the gateway into tBTC providing the redeemer's address. The gateway would lock wthBTC and send tBTC to the redeemer. 
2. Mint to address B on L2 providing the redeemer's address as extra info. B would be the L2 gateway and it would lock whtBTC and send tBTC to the address from the extra info provided.
3. Update Wormhole Portal to execute two transactions during the redemption: Option 1 or Option 2 @conorpp suggested.

The first way is the most user-friendly but I do not know how complicated it is on the Wormhole's side.